pytests/src/enums.rs

```diff
@@ -0,0 +1,50 @@
+use pyo3::{pyclass, pyfunction, pymodule, types::PyModule, wrap_pyfunction, PyResult, Python};
+
+#[pymodule]
+pub fn enums(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
+    m.add_class::<SimpleEnum>()?;
+    m.add_class::<ComplexEnum>()?;
+    m.add_wrapped(wrap_pyfunction!(do_simple_stuff))?;
+    m.add_wrapped(wrap_pyfunction!(do_complex_stuff))?;
+    Ok(())
+}
+
+#[pyclass]
+pub enum SimpleEnum {
+    Sunday,
+    Monday,
+    Tuesday,
+    Wednesday,
+    Thursday,
+    Friday,
+    Saturday,
+}
+
+#[pyfunction]
+pub fn do_simple_stuff(thing: &SimpleEnum) -> SimpleEnum {
+    match thing {
+        SimpleEnum::Sunday => SimpleEnum::Monday,
+        SimpleEnum::Monday => SimpleEnum::Tuesday,
+        SimpleEnum::Tuesday => SimpleEnum::Wednesday,
+        SimpleEnum::Wednesday => SimpleEnum::Thursday,
+        SimpleEnum::Thursday => SimpleEnum::Friday,
+        SimpleEnum::Friday => SimpleEnum::Saturday,
+        SimpleEnum::Saturday => SimpleEnum::Sunday,
+    }
+}
+
+#[pyclass]
+pub enum ComplexEnum {
+    Int { i: i32 },
```

Comment:
What happens if we add a unit variant like `Empty` here. There's three forms:

```rust
    Empty,
    Empty { },
    Empty(),
```

I think the tuple-like case will be similar to the struct-like, and we can solve in a follow-up issue with the rest of the tuple-like variants. For the `Empty {}` struct-like, I think this will already behave as I'd expect (create a new variant subclass with no fields).

But what about the `Empty` identifier? Should this create a new variant too, or should it create a constant like the `SimpleEnum` cases?

I think it has a nice consistency if it works like a `SimpleEnum` variant, but I can also see reasons why that might be confusing. If we are feeling conservative, one option is to disallow `Empty` in `ComplexEnum` for now, track this as a known issue, and invite users to propose the right design later.

Reply:
I've added a test case using `test_compile_error` to document that tuple-like variants are not currently supported.

You're correct; the struct-like `Empty {}` variant corresponds to a PyClass with no fields. I've extended the `pytests` to document this.

My stance is that a unit-like `Empty` variant should behave identically to a struct-like `Empty {}` variant as far as the Python side is concerned. (And that a tuple-like `Empty()` variant should be identical to them as well, when support is added.) This is what is currently implemented, so I've extended the `pytests` to document this.

The alternative, as you suggest, is to make a unit-like `Empty` variant correspond to a constant value on the Python side. This would require special-casing in multiple places in the generator. Because it seems to have little benefit in terms of functionality, apart from removing the `()` in expressions and patterns, I'm against doing this. Please let me know if you think this assessment is incorrect/incomplete.

On the usability side, I think it's also better to _not_ make the unit-like variants in a complex enum behave like the unit-like variants in a simple enum, for the purpose of clearly distinguishing these two types of enums from the user's perspective. A "simple enum" is essentially an int with named values (like `enum` in C), whereas a "complex enum" is a tagged union.

Reply:
My concern is based around silent behaviour changes. Consider the following enum:

```rust
enum Color {
    Red,
    Blue,
    Green,
}
```

In current PyO3, this is already supported, and creates a single type with three constants `Color.Red` etc. which are instances of that type.

Now, add a fourth non-unit variant:

```rust
enum Color {
    Red,
    Blue,
    Green,
    Custom { r: u8, g: u8, b: u8 },
}
```

If we follow the line that `Red` is equivalent to `Red { }`, then just by adding the non-unit variant all the other enums have also changed behaviour. Now to get an instance of the red variant, a Python user must write `Color.Red()`, as `Color.Red` now instead refers to the variant type.

I think it's very likely that users will trip up on this while refactoring code. This makes me feel that the right thing to do is either:
- make sure that `Color.Red` does not change behaviour (as I propose)
- fail to compile and emit an error stating that unit and non-unit variants may not be mixed (which allows us to support this case later)

Reply:
Thanks for describing your concern in more detail, I understand!

I'm in favor of disallowing the mixing of unit and non-unit variants for now, and opening an issue for the remaining work, same as for tuple variants.

To add to your concern about silent behavior changes. Even if we manage to keep the instantiation and pattern syntax consistent across simple and complex enums, there will remain major differences in semantics. Simple enums can be coerced into integers, which is not meaningful for complex enums.

I'm going to make some changes to both disallow unit variants and ensure the code is in a good shape for future extension work.

Reply:
Yes, the conversion also is a problem, so I agree that disallowing unit variants in complex enums makes sense for now.